homeassistant/components/evohome.py

```diff
@@ -0,0 +1,146 @@
+https://home-assistant.io/components/climate.evohome/
+"""
+
+# Glossary:
+# TCS - temperature control system (a.k.a. Controller, Parent), which can
+# have up to 13 Children:
+#   0-12 Heating zones (a.k.a. Zone), and
+#   0-1 DHW controller, (a.k.a. Boiler)
+
+import logging
+import requests
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_USERNAME,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    HTTP_BAD_REQUEST
+)
+
+from homeassistant.exceptions import PlatformNotReady
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.discovery import load_platform
+
+REQUIREMENTS = ['evohomeclient==0.2.7']
+# If ever > 0.2.7, re-check the work-around wrapper is still required when
+# instantiating the client, below.
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'evohome'
+DATA_EVOHOME = 'data_' + DOMAIN
+
+CONF_LOCATION_IDX = 'location_idx'
+MAX_TEMP = 28
+MIN_TEMP = 5
+SCAN_INTERVAL_DEFAULT = 180
+SCAN_INTERVAL_MAX = 300
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_LOCATION_IDX, default=0): cv.positive_int,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+# These are used to help prevent E501 (line too long) violations.
+GWS = 'gateways'
+TCS = 'temperatureControlSystems'
+
+
+def setup(hass, config):
+    """Create a Honeywell (EMEA/EU) evohome CH/DHW system.
+
+    One controller with 0+ heating zones (e.g. TRVs, relays) and, optionally, a
+    DHW controller.  Does not work for US-based systems.
+    """
+    domain_data = hass.data[DATA_EVOHOME] = {}
+    domain_data['timers'] = {}
+
+    domain_data['params'] = dict(config[DOMAIN])
+    domain_data['params'][CONF_SCAN_INTERVAL] = SCAN_INTERVAL_DEFAULT
+
+    from evohomeclient2 import EvohomeClient
+
+    _LOGGER.debug("setup(): API call [4 request(s)]: client.__init__()...")
+
+    try:
+        # There's a bug in evohomeclient2 v0.2.7: the client.__init__() sets
+        # the root loglevel when EvohomeClient(debug=?), so remember it now...
+        log_level = logging.getLogger().getEffectiveLevel()
+
+        client = EvohomeClient(
+            domain_data['params'][CONF_USERNAME],
+            domain_data['params'][CONF_PASSWORD],
+            debug=False
+        )
+        # ...then restore it to what it was before instantiating the client
+        logging.getLogger().setLevel(log_level)
+
+    except requests.exceptions.HTTPError as err:
+        if err.response.status_code == HTTP_BAD_REQUEST:
+            _LOGGER.error(
+                "Failed to establish a connection with evohome web servers, "
+                "Check your username (%s), and password are correct."
+                "Unable to continue. Resolve any errors and restart HA.",
+                domain_data['params'][CONF_USERNAME]
+            )
+            return False  # unable to continue
+
+        raise PlatformNotReady(err)
```

Comment:
Remove this. We can't use this in the component.

Reply:
Should we return False in this case too?

Reply:
I'm inclined to simply raise.  If I do `return False`, then I should add `_LOGGING.error()`.  Which do you think is better...

Reply:
We don't raise. Log error and return False.

Reply:
I only want to handle `HTTP_BAD_REQUEST`, because it is a mis-configuration by the user, and the TraceBack can be a little cryptic (i.e. it is not HTTP_UNAUTHORIZED, or 'bad credentials'). 
 That's why there is an `except` there.  

If I didn't have that `except`, then an exception would occur via the client library.  I am not raising, merely allowing exceptions that I don't want to handle do their natural thing.

I would like: 
```python
    except requests.exceptions.HTTPError as err:
        if err.response.status_code == HTTP_BAD_REQUEST:
            _LOGGER.error(
                "Failed to establish a connection with evohome web servers, "
                "Check your username (%s), and password are correct."
                "Unable to continue. Resolve any errors and restart HA.",
                domain_data['params'][CONF_USERNAME]
            )
            return False  # unable to continue

        raise  # we dont handle any other HTTPErrors
```

Reply:
FYI, here is what `homeassistant.log` looks like when the above scenario occurs, a `503` in this case, and with my preferred solution:
```
2018-09-25 09:58:16 DEBUG (SyncWorker_12) [homeassistant.components.evohome] setup(): API call [4 request(s)]: client.__init__()...
   ...
2018-09-25 09:58:16 ERROR (MainThread) [homeassistant.setup] Error during setup of component evohome
Traceback (most recent call last):
  File "/home/dbonnes/home-assistant/homeassistant/setup.py", line 148, in _async_setup_component
    component.setup, hass, processed_config)  # type: ignore
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/dbonnes/home-assistant/homeassistant/components/evohome.py", line 82, in setup
    debug=False
  File "/home/dbonnes/home-assistant/lib/python3.6/site-packages/evohomeclient2/__init__.py", line 16, in __init__
    self._login()
  File "/home/dbonnes/home-assistant/lib/python3.6/site-packages/evohomeclient2/__init__.py", line 82, in _login
    self._basic_login()
  File "/home/dbonnes/home-assistant/lib/python3.6/site-packages/evohomeclient2/__init__.py", line 70, in _basic_login
    r.raise_for_status()
  File "/home/dbonnes/home-assistant/lib/python3.6/site-packages/requests/models.py", line 939, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 503 Server Error: Service Unavailable for url: https://tccna.honeywell.com/Auth/OAuth/Token
```

Reply:
Assume this is OK.